content/blog/2020-11-11-self/index.md

```diff
@@ -0,0 +1,155 @@
+
+### Customized Compilation
+
+To illustrate this optimization, the authors have us consider a `min` method which calls a `less-than` method. Both the `min` and `less-than` methods are potentially called from many types of objects. A traditional Smalltalk JIT compiler would translate `min` into machine code when it's called on an integer, and reuse the compiled method when called on a floating point. This means `min` needs to be compiled in a type-agnostic way, which prevents Message Inlining (which we'll see shortly) of `less-than`.
+
+However, with customized compilation, an integer-specific version of `min` is compiled when it's called on an integer, and a float-specific version is compiled when it's called on a float. This allows the `less-than` message-passing to be inlined...
+
+### Message Inlining
+
+Remember, the idea of message passing is that the receiver searches itself and its parents for the required slot name, and does something depending on the type of the slot (data, assignment slot, method, block). However, if the type of the receiver is known at compile time (remember what "compile time" means for a JIT), and the message is also known at compile time, the compiler can search for the relevant slot itself so that this dynamic dispatch doesn't have to occur at run-time.
+
+### Primitive Inlining
+
+This optimization inlines calls to builtin primitive operations such as arithmetic or comparison. It also constant-folds the calls if the arguments are known at compile time. In many cases, Customized Compilation enables Primitive Inlining because the types of the arguments are known statically.
+
+### Message Splitting
+
+This is an optimization that occurs at merge points in the control flow graph of a method being compiled. Say that basic blocks A and B both connect to C, and that C passes a message to some object `o`. `o`'s type may depend on whether the A or B path was taken. The optimization is to pass the message to `o` at the end of A and B instead of in C. This makes `o`'s type knowable at both points and enables Message Inlining.
+
+### Type Prediction
+
+This compiler optimization builds upon Message Splitting. The motivation here seems to be mostly based on Self's operator overloading. The idea is that a message like `less-than` will often be passed to integers, floats, and strings because the language provides this behavior, but in rare cases, this message could be passed to arbitrary objects. Then, any time the compiler sees a `less-than` message passed to an object `o`, it inserts branches that check if `o` is an integer, a float, a string, or something else. Then it Message Splits the single message-pass to the ends of these four branches. This allows Message Inlining in the common case and actual message passing in the worst case.
+
+## Supporting the Programming Environment
```

Comment:
Maybe a sentence or two to give context for why we are talking about a programming environment all of a sudden?